### PR TITLE
WIP: New lint [`manual_position`]

### LINKED ISSUE #2282: Iterators: suggest `position()` instead of `enumerate() + find() + .0`
This is sort of the opposite of #456. I recently wrote `let idx = v.iter().enumerate().find(|&(_, &ch)| ch == '|').unwrap().0`, when I should have written `let idx = v.iter().position(|&ch| ch == '|').unwrap()`. I'd have to think a little on the exact rule that should be followed here, but it would be nice if clippy could catch this pattern.

### PR DESCRIPTION
**Pull request by GabrielBFern**
Add a new lint to detect manual implementation of position(_) using enumerate(_).find(_)

changelog: Add new [`manual_position`] lint

fixes #2282

I need help handling this case when I need to change the deref on the body of the closure.
```rust
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().enumerate().find(|(_, ch)|**ch == 3).unwrap().0;
    // Right now, the lint is suggesting this (which does not compile):
    let _ = v.iter().position(|ch|**ch == 3).unwrap();
    // but i think i need to make like this:
    let _ = v.iter().position(|ch|*ch == 3).unwrap();
    // or this:
    let _ = v.iter().position(|&ch|ch == 3).unwrap();
```

### CODE REVIEW

**Comment by GabrielBFern** [comment]
I found this [example](https://github.com/rust-lang/rust-clippy/blob/4c1d05cfa193012eb197119b4614dd979d7f6004/clippy_lints/src/methods/search_is_some.rs#L37-L58). 
I think I need something similar, but I don't know how to create an expr from the closure suggestion I made to pass it to the `deref_closure_args` function. Or if I need to pass the original expr first to `deref_closure_args`, in this case, I cannot separate the arguments and the body of the closure.

**Comment by llogiq** [comment]
First, check whether the closure input pattern already does a deref (`|&a| ..`). If so, we just remove the deref. Otherwise, we can either avoid linting, lint without suggestion or walk through the closure body adding derefs to any `ExprPath` matching the input.

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -4523,16 +4546,27 @@ impl Methods {
...
+                                recv2,
+                                iter_overeager_cloned::Op::FixClosure(name, arg),
+                                false,
+                            );
+                        },
+                        Some(("enumerate", _, [], _, span2)) => {
+                            manual_position::check(cx, expr, arg, span2, false);
+                        },
+                        Some(("rev", recv2, [], _, _)) => {
+                            if let Some(("enumerate", _, [], _, span2)) = method_call(recv2) {
```
Comment (J-ZhengLi) [inline clippy_lints/src/methods/mod.rs:4565]: it's very unlikely, but what happens if someone wrote `iter().rev().rev().enumerate().find()` 
or `iter().rev().enumerate().rev().find()`? 😄

File: tests/ui/manual_position.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    // let _ = v.iter().position(|ch|**ch == 3).unwrap();
+    // but need to be:
+    // let _ = v.iter().position(|ch|*ch == 3).unwrap();
+    // or
+    // let _ = v.iter().position(|&ch|ch == 3).unwrap();
+}
+
+fn rposition_test() {
+    let v = [1, 2, 3, 5, 3, 6];
+    let _ = v.iter().enumerate().rev().find(|(_, &ch)| ch == 3).unwrap().0;
```
Comment (J-ZhengLi) [inline tests/ui/manual_position.rs:72]: let's test it with more "rev"s!!! 😂

File: clippy_lints/src/methods/manual_position.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, arg: &Expr<'_>, start_span: Span, rev: bool) {
+    if let ExprKind::Closure(c) = arg.kind
+        && matches!(c.kind, ClosureKind::Closure)
+        && let typeck = cx.typeck_results()
+        && let Some(fn_id) = typeck.type_dependent_def_id(expr.hir_id)
+        && (is_diag_trait_item(cx, fn_id, sym::Iterator))
+        && let body = cx.tcx.hir().body(c.body)
+        && let [param] = body.params
+        && let parent = cx.tcx.hir().parent_iter(expr.hir_id)
+        && let Some(usage) = parse_usage(cx, parent)
```
Comment (J-ZhengLi) [inline clippy_lints/src/methods/manual_position.rs:34]: Maybe we don't need to give a "usage suggestion" at all, I mean, don't include the span of this call's parent.

for example, we can just use the span of this method call in:

```rust
let _ = seq.iter().enumerate().find(...).unwrap()
//                 ^^^^^^^^^^^^^^^^^^^^^
```

Plus, there are many way to call after an `Option`, such as `unwrap_or_default`, `unwrap_or_else`, `or_else`, `or`, etc. Parsing any missed cases will cause FN.

**Review by J-ZhengLi — COMMENTED** [review]
Came here for triaging, stayed for reviewing...

fwiw, `manual_position` sounds a bit vague, how about calling it `manual_iter_position`?

**Comment by J-ZhengLi** [comment]
Now here comes the triaging part, lol...



Hey @GabrielBFern, this is a ping from triage. There hasn't been any activity for some time, you're always welcome to ask questions (there's a [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) if you're interested). It's also fine if you don't have the time right now. We just want to check in and make sure that you're not stuck.

@rustbot author

**Comment by xFrednet** [comment]
Hey this is triage, I'm closing this due to inactivity. If you want to continue this implementation, you're welcome to create a new PR. Thank you for the time, you already put into this!

Interested parties are welcome to pick this implementation up as well :)

@rustbot label +S-inactive-closed -S-waiting-on-author -S-waiting-on-review

**Pull request closed without merging**